### PR TITLE
Adding ability to unfurl links

### LINKED ISSUE #526: support link unfurling
Currently, URLs within the `message` are not "unfurled" in slack parlance. Is it possible to fix this from within the `slackSend` source code? If so, then consider this a feature request for the same.

### PR DESCRIPTION
**Pull request by EricBartusch**
Added a new optional boolean parameter: `sendAsText` as I don't know of any way to have Slack unfurl links found in attachments. Not including the parameter or setting as `false` returns the same functionality as before.  Setting it to `true` causes the plugin to send a message as text instead of in an attachment and because I added `unfurl_links=true` and `unfurl_media=true` to the API calls, links will be automatically unfurled in Slack.

I've tested this both as a bot user and as the default Jenkins user.

Fixes #526

### CODE REVIEW

**Comment by casz** [comment]
@EricBartusch 
You might want to set up an email alias on your GitHub account 🥇 
Or look into fixing your git user.email config 😅 
to get proper credit for your commits 🏆 
https://help.github.com/articles/adding-an-email-address-to-your-github-account/
https://help.github.com/articles/setting-your-commit-email-address-in-git
then to amend your commits: https://stackoverflow.com/a/3042512

Seems your git email changed here: https://github.com/jenkinsci/slack-plugin/commit/c1e320652dd11c9935a2088eb228e94969f215dc.patch

It's fine on this commit https://github.com/jenkinsci/slack-plugin/commit/95d60ee2cb7e213b491eb71e17510e32c229d3b6.patch

**Comment by EricBartusch** [comment]
The pains of switching between a work PC and my home PC!

**Comment by casz** [comment]
@EricBartusch This is totally solvable even when switching: 
https://github.com/casz/dotfiles/blob/f37e0bddf17d9949316e1eeb832486283c1636d2/git/gitconfig#L59-L66

**Comment by EricBartusch** [comment]
I realized that I didn't add this functionality for freestyle jobs.  Should I close the PR until I add that?

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 1]
```diff
@@ -164,15 +167,21 @@ public boolean publish(String message, JSONArray attachments, String color) {
...
+                json.put("unfurl_media", "true");
 
                 nvps.add(new BasicNameValuePair("payload", json.toString()));
             } else {
                 url = "https://slack.com/api/chat.postMessage?token=" + populatedToken +
                         "&channel=" + roomId.replace("#", "") +
                         "&link_names=1" +
-                        "&as_user=true";
+                        "&as_user=true" +
+                        "&unfurl_links=true" +
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: After seeing this again, I was like WAIT A SECOND.. :laughing: 
You do know this very same endpoint accepts JSON?

https://api.slack.com/methods/chat.postMessage
> Accepted content types: application/x-www-form-urlencoded, application/json

Why are we building a URL when we could build JSON?

**Comment by timja** [comment]
I don't mind if this doesn't work on freestyle jobs, pipeline jobs are a lot more flexible with this plugin...

**Comment by EricBartusch** [comment]
I'll keep this out of freestyle jobs.  What are your thoughts on having it globally configured? I can add that in if it's worthwhile.

**Comment by timja** [comment]
Yeah worth having it globally

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 2]
```diff
@@ -134,7 +134,7 @@ public void completed(AbstractBuild r) {
         if (null != previousBuild) {
             do {
                 previousBuild = previousBuild.getPreviousCompletedBuild();
-            } while ((null != previousBuild && previousBuild.getResult() == Result.ABORTED) || (null != previousBuild && previousBuild.getNumber() == r.getNumber()));
+            } while (null != previousBuild && previousBuild.getResult() == Result.ABORTED);
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: why change this?

Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: I screwed up my branches when I first forked the repo.  I made the above bug fix in master, made a PR, then made a branch for this feature.  I undid that work in the feature branch and forgot to add it back in.

TL;DR: Whoops

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 3]
```diff
@@ -96,6 +98,7 @@ public StandardSlackService(StandardSlackServiceBuilder standardSlackServiceBuil
         }
         this.roomIds = standardSlackServiceBuilder.roomId.split("[,; ]+");
         this.replyBroadcast = standardSlackServiceBuilder.replyBroadcast;
+        this.sendAsText = standardSlackServiceBuilder.sendAsText;
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: Seems unused here? Are we missing an if statement around unfurl_links/unfurl_media?

Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/StandardSlackService.java]: If that line is removed, the `sendAsText` attribute is never applied.  An if statement around unfurl_links/unfurl_media would make things more clear, but isn't necessary.  If messages are being sent as attachments, which is the default, those parameters are benign. I can see the case for adding the if around them though.

Comment (casz) [reply src/main/java/jenkins/plugins/slack/StandardSlackService.java]: Then what's the sendAsText for?

File: src/main/java/jenkins/plugins/slack/workflow/SlackSendStep.java [hunk 4]
```diff
@@ -286,10 +298,11 @@ protected SlackResponse run() throws Exception {
                 return null;
             }
 
-            SlackService slackService = getSlackService(
-                    baseUrl, teamDomain, botUser, channel, step.replyBroadcast, iconEmoji, username, populatedToken);
+            SlackService slackService = getSlackService(baseUrl, teamDomain, botUser, channel, step.replyBroadcast, sendAsText, iconEmoji, username, populatedToken);
             final boolean publishSuccess;
-            if (step.attachments != null) {
+            if (sendAsText) {
```
Comment (EricBartusch) [inline src/main/java/jenkins/plugins/slack/workflow/SlackSendStep.java]: @casz sendAsText is used here to skip over creating the message as an attachment

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java (hunk 3 above)
Comment (timja) [reply src/main/java/jenkins/plugins/slack/StandardSlackService.java]: this only seems to be used in slacksendstep? does it need to be here?

File: src/test/java/jenkins/plugins/slack/SlackNotifierTest.java [hunk 5]
```diff
@@ -51,7 +51,7 @@ public void testDoTestConnection() {
         }
         descriptor.setSlackService(slackServiceStub);
         FormValidation result = descriptor
-                .doTestConnection("baseUrl", "teamDomain", "authTokenCredentialId", false, "room", ":+1:", "slack", null);
+                .doTestConnection("baseUrl", "teamDomain", "authTokenCredentialId", false, "room",false, ":+1:", "slack", null);
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/SlackNotifierTest.java]:
```suggestion
                .doTestConnection("baseUrl", "teamDomain", "authTokenCredentialId", false, "room", false, ":+1:", "slack", null);
```

**Comment by timja** [comment]
functionality works, links unfurl now, just need the two review comments sorted

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java (hunk 3 above)
Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/StandardSlackService.java]: You are both correct, it's not necessary.

File: src/main/java/jenkins/plugins/slack/StandardSlackServiceBuilder.java [hunk 6]
```diff
@@ -39,6 +40,11 @@ public StandardSlackServiceBuilder withReplyBroadcast(boolean replyBroadcast) {
         return this;
     }
 
+    public StandardSlackServiceBuilder withSendAsText(boolean sendAsText) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/StandardSlackServiceBuilder.java]: this is unused currently

File: src/main/resources/jenkins/plugins/slack/SlackNotifier/global.jelly [hunk 7]
```diff
@@ -25,6 +25,9 @@
     <f:entry title="Is Bot User?" help="/plugin/slack/help-globalConfig-botUser.html">
         <f:checkbox field="botUser" />
     </f:entry>
+    <f:entry title="Send as Text?" help="/plugin/slack/help-globalConfig-sendAsText.html">
+        <f:checkbox field="sendAsText" />
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/SlackNotifier/global.jelly]: this isn't working currently

Comment (EricBartusch) [reply src/main/resources/jenkins/plugins/slack/SlackNotifier/global.jelly]: Are you sure?  It's working fine for me.

Comment (timja) [reply src/main/resources/jenkins/plugins/slack/SlackNotifier/global.jelly]: ah my bad it is working

**Comment by EricBartusch** [comment]
Thanks for working through the issues with this PR and the last. My OOP skills are a little dull to say the least.
